pkg/dataobj/internal/dataset/page.go

```diff
@@ -0,0 +1,104 @@
+package dataset
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash/crc32"
+	"io"
+
+	"github.com/golang/snappy"
+	"github.com/klauspost/compress/zstd"
+
+	"github.com/grafana/loki/v3/pkg/dataobj/internal/metadata/datasetmd"
+)
+
+// Helper types.
+type (
+	// PageData holds the raw data for a page.
+	PageData []byte
+
+	// PageInfo describes a page.
+	PageInfo struct {
+		UncompressedSize int    // UncompressedSize is the size of a page before compression.
+		CompressedSize   int    // CompressedSize is the size of a page after compression.
+		CRC32            uint32 // CRC32 checksum of the page after encoding and compression.
+		RowCount         int    // RowCount is the number of rows in the page, including NULLs.
+
+		Value       datasetmd.ValueType       // Type of values stored in the page.
+		Compression datasetmd.CompressionType // Compression used for values in the page.
+		Encoding    datasetmd.EncodingType    // Encoding used for values in the page.
+		Stats       *datasetmd.Statistics     // Optional statistics for the page.
+	}
+)
+
+// MemPage holds an encoded (and optionally compressed) sequence of [Value]
+// entries of a common type. Use [pageBuilder] to construct pages.
+type MemPage struct {
```

Comment:
I'm reserving `dataset.Page` here for an interface that'll come into play in the next PR.